doc/design/effects_mutability.md

```diff
@@ -0,0 +1,266 @@
+# Effects and mutability
+
+Hemlock statically determines all effects and mutability transitively associated
+with a type, whether direct or indirect.  Internal effects and mutability are
+prohibited in global and message values, though external effects (i.e. outside
+the Hemlock execution environment) are allowed.  Internal effects are lexically
+bound to a particular actor's local heap, and therefore cannot be supported in
+the global heap.  These restrictions are evaluated at compile time, which means
+that the programming model is unaffected by type erasure.  The absence of
+dynamic validation overhead enables fast message passing, as well as fully
+automated migration of compatible long-lived values to the global heap.
+
+Non-parametric types have fixed effects and mutability, whereas parametric types
+also have transitive effects and/or mutability that allow differing
+parametrizations to be any combination of effectless/effectful and/or
+immutable/mutable.  For example, `'a array` is directly immutable, but it may be
+transitively mutable depending on the particulars of the type supplied to the
+`'a` parameter.  Were we to attempt passing `'a array` in a message, the
+compiler would be unable to prove that the message is compatible with the
+aforementioned constraints.  Therefore we must specify transitive constraints in
+the type, e.g. `'a <.  \& array`.  Similarly, finalization is only supported for
+transitively mutable values, e.g. `/& 'a`, because referential transparency of
+immutable values makes finalization ill-defined.
+
+Type syntax is rather involved in its most general form, but most of its
+complexities only come into play with functorized module types.  Following are
+some examples from the Basis library.
+
+```hemlock
+'a list
+'a /r array
+(\& 'k, 'v, 'cmp) map
+(\& 'a, 'cmp) set
+
+val finalize: (/& 'a >-> unit) -> /& 'a >-> unit
+
+type 'a elm =
+| Nil
+| Cons of 'a * ('a, >e) stream
+and ('a, >e) stream = ('a elm, >e) Lazy.t
+val force: ('a, >e) lazy_t >e-> 'a
+```
+
+Parametric type `'a` carries with it effect and mutability parameters, which can
+be explicitly written as `>a /a 'a`.  For example, the set type could be written
+as `(>a /a\& 'a, >cmp /cmp 'cmp) set`.  The `/a` parameter is constrained to
+prohibit mutability, lest key mutation corrupt the set.
+
+Regarding effects versus mutability consider that `'a &array` is a mutable type,
+but actual mutation is an effect.
+
+```hemlock
+val length: 'a /_ array -> usize
+val set: 'a -> usize -> 'a !&array -> unit
```

Comment:
I don't think I ever fully grasped what counts as an internal read-write effect. What I wonder is if mutating a functions input counts as an internal write effect.
```ocaml
let foo arr = begin
  let bar arr () = begin
    Array.set_inplace 0 42 arr
  end in
  bar arr
```
Would the type signature here be
```Ocaml
val foo: uns !&array -> (unit !-> unit)
```
The reason I ask is that I think it might matter that we specify where the mutable input actually gets mutated. Otherwise, that type signature I just showed might be
```Ocaml
val foo: uns !&array -> (unit -> unit)
```
Which clearly has some problems. I think the example you show on this line has the same problem, and I'm wondering if it's basically a typo or if we missed something in the type system.

Also, I'm guessing we'll have global values. Do I have some of these examples correct?

Global read implementation.
```Ocaml
let arr = Array.init 42 f:(i -> i)
let foo () = begin
  arr[0]
end
```
Global read signature.
```Ocaml
val foo: unit .-> uns
```

Global write implementation.
```Ocaml
let arr = Array.init 42 f:(i -> i)
let foo () = begin
  let () = Array.set_inplace 0 1 arr in
  arr[0]
end
```
Global write signature.
```Ocaml
val foo: unit !-> uns
```

Reply:
I added a section regarding partial application and (not) losing track of effects.  In short, the first `foo` example you gave does need a `!->` arrow in the resulting closure, and there's a principled way to always get this right.

I also rearranged things and added a gentler intro to effects and mutability.  Let me know if it still needs help to be straightforward reading.